internal/locCluster.ecl

```diff
@@ -0,0 +1,314 @@
+
+            if(p->wi != ds[i]->wi) continue;
+
+            double dist;
+
+            if(distanceFunc.compare("cosine")==0)
+                dist = cosine(p->fields, ds[i]->fields);
+            else if(distanceFunc.compare("minkowski")==0)
+                dist = minkowski(p->fields, ds[i]->fields, dist_params[0]);
+            else if(distanceFunc.compare("manhattan")==0)
+                dist = manhattan(p->fields, ds[i]->fields);
+            else if(distanceFunc.compare("haversine")==0)
+                dist = haversine(p->fields, ds[i]->fields);
+            else if(distanceFunc.compare("chebyshev")==0)
+                dist = chebyshev(p->fields, ds[i]->fields);
+            else
+                dist = euclidian(p->fields, ds[i]->fields);
+
+            if(dist <= eps) ret.push_back(ds[i]);
+        }
+        return ret;
+    }
+
+    // function: find returns the ultimate parent of the node in tree.
+    // input: node * data pointer
+    // output: pointer to parent of tree.
+    node* find(node *p){
+        if(p->parent == NULL || p->parent == p){
+            return p;
+        } else {
+            return find(p->parent);
+        }
+    }
+    // function: Union merges the trees a and b based on disjoint sets
+    // input : Two trees(a and b) to merge
+    // output: void.
+    void Union(vector<node*> ds, node* a, node* b){
+        node* x = find(a);
+        node* y = find(b);
+        if(x->ifCore && !y->ifCore){
+            y->parent = x;
+            return;
+        } else if (!x->ifCore && y->ifCore){
+            x->parent = y;
+            return;
+        }
+        if(x == y) return;
+        else if(x->id > y->id) y->parent = x;
+        else x->parent = y;
+    }
+    //function: dbscan returns a dataset with parentid's in each local node.
+    //input: Vector ds, distance eps and number od min points.
+    void dbscan(vector<node*> ds, double eps, uint64_t minpts) {
+        for(uint64_t i=0; i<ds.size(); ++i){
+
+            if(!ds[i]->ifLocal) continue;
+
+            ds[i]->isModified = true;
+            vector<node*> neighs = getNeighbors(ds,ds[i],eps);
+
+            if(neighs.size() >= minpts){
+                ds[i]->ifCore = true;
+                ds[i]->parent = NULL;
+                for(uint64_t n=0; n < neighs.size(); ++n){
+                    neighs[n]->isModified = true;
+                    if(neighs[n]->ifLocal){
+                        if(neighs[n]->ifCore){
+                            Union(ds,ds[i],neighs[n]);
+                        } else {
+                            if(neighs[n]->isVisited) continue;
+                            neighs[n]->isVisited = true;
+                            Union(ds,ds[i],neighs[n]);
+                        }
+                    } else {
+                        if(!neighs[n]->ifCore){
+                            if(getNeighbors(ds,neighs[n],eps).size()>=minpts){
+                                neighs[n]->ifCore = true;
```

Comment:
If this is a common code path it might be worth considering adding a new function derived from getNeighbors that checks if it has at least n neighbours.  If would avoid returning the vector, and it would allow the loop to be short-circuited.

Reply:
If it's okay, I prefer to updating it in the next stage.

Reply:
I agree.